[[ LCB ]] Improve tokenisation of numbers and add hex and bin literals

runrevmark opened the pull request:
This patch improves the tokenisation of numbers in lc-compile and adds
support for hexadecimal and binary integer literals.

The compiler will now complain about any invalid suffixes on numeric
literals, meaning that it is not possible to accidentally elide a
numeric literal with an identifier.

Hexadecimal literals are specified using the prefix "0x".

Binary literals are specified using the prefix "0b".

Real literals have been made more flexible and accurate. In particular,
leading zeros are now checked for in all but the fractional part, and
considering a real number to be of the form:

```
Int DecPoint Frac Exp
```

The allowed forms are:

```
 Int+DecPoint
 Int+Exp
 Int+DecPoint+Exp
 Int+DecPoint+Frac
 Int+DecPoint+Frac+Exp
 DecPoint+Frac
 DecPoint+Frac+Exp
```

All other combinations will be flagged as a 'malformed real literal'
error.

Note: As the regex for integers now intersects with that for reals
(e.g. 1e1 could be an integer with suffix, or a real) rules for both
DOUBLE and INTEGER literals are now in DOUBLE_LITERAL.t with the DOUBLE
rules first. This ensures DOUBLE_LITERAL matches in preference to
INTEGER_LITERAL.

runrevmark committed:
[[ CompilerTestRunner ]] Emit source with line numbers in verbose mode

peter-b commented on toolchain/lc-compile/foo.c:
```diff
@@ -0,0 +1,14 @@
+        printf("%ul %d %d %d\n", value, errno, ERANGE, *end);
+	return 0;
+}
+
```
Did you mean to commit this file?

runrevmark replied:
No I didn't :)

peter-b commented on toolchain/lc-compile/src/DOUBLE_LITERAL.t:
```diff
@@ -1,6 +1,45 @@
+	/* This contains both INTEGER_LITERAL and DOUBLE_LITERAL rules */
+	/* This ensures that they appear in this order, to ensure they match */
+	/* correctly. */
+
```
Is it feasible to rename this file to `NUMERIC_LITERAL.t` and remove `INTEGER_LITERAL.t` entirely?

runrevmark replied:
Unfortunately not - I tried that but it makes reflex choke - it expects a .t file for each token.

peter-b commented on toolchain/lc-compile/src/literal.c:
```diff
@@ -87,6 +161,166 @@ int MakeIntegerLiteral(const char *p_token, long *r_literal)
+		  (t_found_point && t_found_frac && t_found_exp)))
+	{
+		return 0;
+	}*/
```
Can this code block be removed, or does it need to be retained for explanatory purposes?

peter-b commented:
@runrevmark I think we need to include release notes for this change.

runrevmark committed:
[[ LCB ]] Improve tokenisation of numbers and add hex and bin literals
This patch improves the tokenisation of numbers in lc-compile and adds
support for hexadecimal and binary integer literals.

The compiler will now complain about any invalid suffixes on numeric
literals, meaning that it is not possible to accidentally elide a
numeric literal with an identifier.

Hexadecimal literals are specified using the prefix "0x".

Binary literals are specified using the prefix "0b".

Real literals have been made more flexible and accurate. In particular,
leading zeros are now checked for in all but the fractional part, and
considering a real number to be of the form:
     Int DecPoint Frac Exp
The allowed forms are:
     Int+DecPoint
     Int+Exp
     Int+DecPoint+Exp
     Int+DecPoint+Frac
     Int+DecPoint+Frac+Exp
     DecPoint+Frac
     DecPoint+Frac+Exp
All other combinations will be flagged as a 'malformed real literal'
error.

Note: As the regex for integers now intersects with that for reals
(e.g. 1e1 could be an integer with suffix, or a real) rules for both
DOUBLE and INTEGER literals are now in DOUBLE_LITERAL.t with the DOUBLE
rules first. This ensures DOUBLE_LITERAL matches in preference to
INTEGER_LITERAL.

runrevmark commented:
@peter-b: Release note added :)

peter-b commented on docs/lcb/notes/feature-hex_bin_literals.md:
```diff
@@ -0,0 +1,20 @@
+* Base 2 (binary) integer literals can now be specified by using a "0b" prefix, e.g.
+
+      0b0000
+      0b1010
```
Due to a quirk of Markdown, I believe this code block and the code block in the following bullet point will only get successfully parsed as code blocks if you wrap the first paragraph of bullet point, e.g.:

``` markdown
* Base 2 (binary) integer literals can now be specified by using a "0b"
  prefix, e.g.

      0b0000
      0b1010
```

Yes, Markdown is horrible.

**Edit**: Behold! You can see the problem if you [view this file in GitHub's Markdown viewer](https://github.com/runrevmark/livecode/blob/lcb-hex_binary_literals/docs/lcb/notes/feature-hex_bin_literals.md).

peter-b commented on toolchain/lc-compile/src/literal.c:
```diff
@@ -87,6 +161,155 @@ int MakeIntegerLiteral(const char *p_token, long *r_literal)
+	#define I 8
+	#define P 4
+	#define F 2
+	#define E 1
```
Please move these definitions up the function and use them when computing `t_found_bits`, e.g.

``` C
t_found_bits =
    (t_found_int ? I : 0) |
    (t_found_point ? P : 0) |
    (t_found_frac ? F : 0) |
    (t_found_exp ? E : 0);
```

livecodeali replied:
Are we still discouraging use of the ternary operator?

runrevmark replied:
Well the final patch didn't have them in ;)

peter-b commented on toolchain/lc-compile/src/report.c:
```diff
@@ -237,6 +237,8 @@ DEFINE_ERROR_I(UnableToFindImportedModule, "Unable to find imported module '%s'"
 DEFINE_ERROR_S(MalformedToken, "Illegal token '%s'");
 DEFINE_ERROR_S(MalformedEscapedString, "Illegal escape in string '%s'");
 DEFINE_ERROR(MalformedSyntax, "Syntax error");
+DEFINE_ERROR(InvalidIntegerLiteral, "Malformed integer literal");
```
Definitely out of the scope of this PR, but it would be quite nice to have an explicit "Integer literal too large" error message that's distinct from "Malformed integer literal" since they're not quite the same thing.

runrevmark replied:
To be fair its not out of scope really - I'll add that now.

peter-b commented:
Nice feature to have; I particularly appreciate the comprehensive tests 💯 

@livecode-vulcan review ok 2c4cf57

livecode-vulcan commented:
💙 review by @peter-b ok 2c4cf57e8ec66633b01d5682760283d0b9517897

livecode-vulcan commented:
😞 test failure 2c4cf57e8ec66633b01d5682760283d0b9517897
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/200)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/204)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/200)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/200)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/204)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/200)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/200)

peter-b commented:
@livecode-vulcan retry 2c4cf57

livecode-vulcan commented:
😞 test failure 2c4cf57e8ec66633b01d5682760283d0b9517897
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/215)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/219)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/215)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/215)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/219)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/215)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/215)

runrevmark commented:
@peter-b: Could you re-review - windows didn't like my non-C89 functions :(

peter-b commented:
@livecode-vulcan review ok b608944

livecode-vulcan commented:
💙 review by @peter-b ok b6089445e024da8639aa76aeafcdfe1803cd62ca

livecode-vulcan commented:
😎 test success b6089445e024da8639aa76aeafcdfe1803cd62ca
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/216)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/220)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/216)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/216)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/220)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/216)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/216)
